File: scripts/convert-tsr-corpus.py
#!/usr/bin/env python3
"""Convert tree-sitter-r test corpus files into the project fixture format.

Reads every ``corpus/tree-sitter-r/*.txt`` file, splits it into the tree-sitter
test sections (name / source / expected S-expression), and emits one project
fixture file per corpus file at ``crates/syntax/tests/tsr/<stem>.R.test``. The
original expected S-expression of each section is written alongside to
``corpus/tree-sitter-r-expected/<stem>/<slug>.sexp`` (gitignored) so the parser's
output can later be adjudicated against tree-sitter's.

Tree-sitter corpus grammar (with the robustness rules this converter relies on):

  * A section header is a line of only ``=`` (>= 3), then one or more name /
    attribute lines (none of which is an all-``=`` line), then another line of
    only ``=`` (>= 3). Attribute lines start with ``:`` (e.g. ``:skip``,
    ``:error``, ``:language(...)``).
  * The source code follows the header, then a divider line of only ``-``
    (>= 3), then the expected S-expression, which runs until the next header.
  * R source can itself contain lines of ``-`` or ``=``. We disambiguate with two
    facts that always hold for real corpora: a header is only a header when a
    closing ``=`` fence follows the name block, and the expected S-expression
    never contains a standalone all-``-`` line, so the *last* ``-`` divider in a
    section body is the true source/expected split.

The expectation body we emit is intentionally empty (a single blank line): these
fixtures stage tree-sitter's cases for the hand-written parser; the parser's
golden trees are filled in later. Sections marked ``:skip`` are dropped. Sections
whose expected S-expression contains ``(ERROR`` or ``(MISSING`` (recovery tests)
are emitted with a ``_ts_error`` case-name suffix.
"""

from __future__ import annotations

import os
import sys
from pathlib import Path

REPO_ROOT = Path(__file__).resolve().parent.parent
CORPUS_DIR = REPO_ROOT / "corpus" / "tree-sitter-r"
FIXTURE_DIR = REPO_ROOT / "crates" / "syntax" / "tests" / "tsr"
EXPECTED_DIR = REPO_ROOT / "corpus" / "tree-sitter-r-expected"


def is_fence(line: str, char: str) -> bool:
    stripped = line.rstrip()
    return len(stripped) >= 3 and all(character == char for character in stripped)


def is_equals(line: str) -> bool:
    return is_fence(line, "=")


def is_dash(line: str) -> bool:
    return is_fence(line, "-")


def try_header(lines: list[str], index: int) -> tuple[int, list[str]] | None:
    """If a section header opens at ``lines[index]``, return the index of its
    closing ``=`` fence and the header content lines between the fences."""
    if index >= len(lines) or not is_equals(lines[index]):
        return None
    cursor = index + 1
    content: list[str] = []
    while cursor < len(lines) and not is_equals(lines[cursor]):
        content.append(lines[cursor])
        cursor += 1
    if cursor >= len(lines):
        return None  # no closing fence: not a header
    if not any(line.strip() for line in content):
        return None  # empty header block: not a header
    return cursor, content


def slugify(name: str) -> str:
    slug = "".join(character.lower() if character.isalnum() else "_" for character in name)
    slug = slug.strip("_")
    while "__" in slug:
        slug = slug.replace("__", "_")
    return slug or "case"


def strip_blank_edges(lines: list[str]) -> list[str]:
    start = 0
    end = len(lines)
    while start < end and not lines[start].strip():
        start += 1
    while end > start and not lines[end - 1].strip():
        end -= 1
    return lines[start:end]


def parse_sections(text: str) -> list[tuple[str, list[str], list[str], list[str]]]:
    """Return ``(name, attributes, source_lines, expected_lines)`` per section."""
    lines = text.replace("\r\n", "\n").replace("\r", "\n").split("\n")
    sections: list[tuple[str, list[str], list[str], list[str]]] = []

    index = 0
    while index < len(lines):
        header = try_header(lines, index)
        if header is None:
            index += 1  # stray content before the first header (or a blank line)
            continue
        close_index, content = header
        name = " ".join(
            line.strip() for line in content if not line.strip().startswith(":")
        ).strip()
        attributes = [line.strip() for line in content if line.strip().startswith(":")]

        body_start = close_index + 1
        scan = body_start
        next_header = len(lines)
        while scan < len(lines):
            if try_header(lines, scan) is not None:
                next_header = scan
                break
            scan += 1
        body = lines[body_start:next_header]

        dash_positions = [position for position, line in enumerate(body) if is_dash(line)]
        if dash_positions:
            divider = dash_positions[-1]
            source = body[:divider]
            expected = body[divider + 1:]
        else:
            source = body
            expected = []

        sections.append((name, attributes, strip_blank_edges(source), strip_blank_edges(expected)))
        index = next_header

    return sections


def convert_file(path: Path) -> tuple[int, int]:
    """Convert one corpus file. Returns ``(emitted, skipped)`` section counts."""
    stem = path.stem
    text = path.read_text(encoding="utf-8", errors="replace")
    sections = parse_sections(text)

    fixture_lines: list[str] = [f"#==== {stem}", ""]
    expected_stem_dir = EXPECTED_DIR / stem
    used_slugs: dict[str, int] = {}
    emitted = 0
    skipped = 0
    collisions: list[str] = []

    for name, attributes, source, expected in sections:
        if any(attribute.startswith(":skip") for attribute in attributes):
            skipped += 1
            continue

        expected_text = "\n".join(expected)
        is_error = (
            "(ERROR" in expected_text
            or "(MISSING" in expected_text
            or any(attribute.startswith(":error") for attribute in attributes)
        )

        slug = slugify(name)
        if is_error:
            slug = f"{slug}_ts_error"
        count = used_slugs.get(slug, 0) + 1
        used_slugs[slug] = count
        if count > 1:
            slug = f"{slug}_{count}"

        for source_line in source:
            if source_line.startswith(("#====", "#----", "#++++")):
                collisions.append(f"{stem}/{slug}: {source_line!r}")

        fixture_lines.append(f"#---- {slug}")
        fixture_lines.extend(source)
        fixture_lines.append("#++++")
        fixture_lines.append("")

        expected_stem_dir.mkdir(parents=True, exist_ok=True)
        (expected_stem_dir / f"{slug}.sexp").write_text(
            expected_text + "\n", encoding="utf-8"
        )
        emitted += 1

    FIXTURE_DIR.mkdir(parents=True, exist_ok=True)
    (FIXTURE_DIR / f"{stem}.R.test").write_text("\n".join(fixture_lines) + "\n", encoding="utf-8")

    if collisions:
        print(f"  WARNING: {len(collisions)} source line(s) collide with fixture directives:")
        for collision in collisions:
            print(f"    {collision}")

    return emitted, skipped


def main() -> int:
    if not CORPUS_DIR.is_dir():
        print(f"corpus not found at {CORPUS_DIR} (run scripts/fetch-corpus.sh first)", file=sys.stderr)
        return 1

    corpus_files = sorted(CORPUS_DIR.glob("*.txt"))
    if not corpus_files:
        print(f"no *.txt files under {CORPUS_DIR}", file=sys.stderr)
        return 1

    total_emitted = 0
    total_skipped = 0
    for path in corpus_files:
        emitted, skipped = convert_file(path)
        total_emitted += emitted
        total_skipped += skipped
        print(f"{path.name}: {emitted} emitted, {skipped} skipped")

    print(
        f"\nconverted {len(corpus_files)} file(s): "
        f"{total_emitted + total_skipped} sections total, "
        f"{total_emitted} emitted, {total_skipped} skipped"
    )
    print(f"fixtures -> {FIXTURE_DIR.relative_to(REPO_ROOT)}")
    print(f"expected sexps -> {EXPECTED_DIR.relative_to(REPO_ROOT)}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/convert-tsr-corpus.rs
#!/usr/bin/env -S cargo +nightly -Zscript
---
[package]
edition = "2024"
---

//! Convert tree-sitter-r test corpus files into the project fixture format.
//! Run with `cargo +nightly -Zscript scripts/convert-tsr-corpus.rs`.
//!
//! Reads every `corpus/tree-sitter-r/*.txt` file, splits it into the tree-sitter
//! test sections (name / source / expected S-expression), and emits one project
//! fixture file per corpus file at `crates/syntax/tests/tsr/<stem>.R.test`. The
//! original expected S-expression of each section is written alongside to
//! `corpus/tree-sitter-r-expected/<stem>/<slug>.sexp` (gitignored) so the parser's
//! output can later be adjudicated against tree-sitter's.
//!
//! Tree-sitter corpus grammar (with the robustness rules this converter relies on):
//!
//! * A section header is a line of only `=` (>= 3), then one or more name /
//!   attribute lines (none of which is an all-`=` line), then another line of
//!   only `=` (>= 3). Attribute lines start with `:` (e.g. `:skip`, `:error`,
//!   `:language(...)`).
//! * The source code follows the header, then a divider line of only `-`
//!   (>= 3), then the expected S-expression, which runs until the next header.
//! * R source can itself contain lines of `-` or `=`. We disambiguate with two
//!   facts that always hold for real corpora: a header is only a header when a
//!   closing `=` fence follows the name block, and the expected S-expression
//!   never contains a standalone all-`-` line, so the *last* `-` divider in a
//!   section body is the true source/expected split.
//!
//! The expectation body we emit is intentionally empty (a single blank line): these
//! fixtures stage tree-sitter's cases for the hand-written parser; the parser's
//! golden trees are filled in later. Sections marked `:skip` are dropped. Sections
//! whose expected S-expression contains `(ERROR` or `(MISSING` (recovery tests)
//! are emitted with a `_ts_error` case-name suffix.

use std::collections::HashMap;
use std::error::Error;
use std::fs;
use std::path::{Path, PathBuf};
use std::process::ExitCode;

const CORPUS_SUBDIR: &str = "corpus/tree-sitter-r";
const FIXTURE_SUBDIR: &str = "crates/syntax/tests/tsr";
const EXPECTED_SUBDIR: &str = "corpus/tree-sitter-r-expected";

fn main() -> ExitCode {
    match run() {
        Ok(code) => code,
        Err(error) => {
            eprintln!("{error}");
            ExitCode::FAILURE
        }
    }
}

fn run() -> Result<ExitCode, Box<dyn Error>> {
    let repo_root = repo_root()?;
    let corpus_dir = repo_root.join(CORPUS_SUBDIR);
    let fixture_dir = repo_root.join(FIXTURE_SUBDIR);
    let expected_dir = repo_root.join(EXPECTED_SUBDIR);

    if !corpus_dir.is_dir() {
        eprintln!(
            "corpus not found at {} (run scripts/fetch-corpus.rs first)",
            corpus_dir.display()
        );
        return Ok(ExitCode::FAILURE);
    }

    let mut corpus_files: Vec<PathBuf> = fs::read_dir(&corpus_dir)?
        .collect::<Result<Vec<_>, _>>()?
        .into_iter()
        .map(|entry| entry.path())
        .filter(|path| path.file_name().is_some_and(|name| name.to_string_lossy().ends_with(".txt")))
        .collect();
    corpus_files.sort();
    if corpus_files.is_empty() {
        eprintln!("no *.txt files under {}", corpus_dir.display());
        return Ok(ExitCode::FAILURE);
    }

    let mut total_emitted = 0;
    let mut total_skipped = 0;
    for path in &corpus_files {
        let (emitted, skipped) = convert_file(path, &fixture_dir, &expected_dir)?;
        total_emitted += emitted;
        total_skipped += skipped;
        let name = path.file_name().unwrap_or_default().to_string_lossy();
        println!("{name}: {emitted} emitted, {skipped} skipped");
    }

    println!(
        "\nconverted {} file(s): {} sections total, {} emitted, {} skipped",
        corpus_files.len(),
        total_emitted + total_skipped,
        total_emitted,
        total_skipped
    );
    println!("fixtures -> {FIXTURE_SUBDIR}");
    println!("expected sexps -> {EXPECTED_SUBDIR}");
    Ok(ExitCode::SUCCESS)
}

/// Convert one corpus file. Returns `(emitted, skipped)` section counts.
fn convert_file(
    path: &Path,
    fixture_dir: &Path,
    expected_dir: &Path,
) -> Result<(usize, usize), Box<dyn Error>> {
    let stem = path
        .file_stem()
        .ok_or_else(|| format!("corpus file has no stem: {}", path.display()))?
        .to_string_lossy()
        .into_owned();
    let text = String::from_utf8_lossy(&fs::read(path)?).into_owned();
    let sections = parse_sections(&text);

    let mut fixture_lines: Vec<String> = vec![format!("#==== {stem}"), String::new()];
    let expected_stem_dir = expected_dir.join(&stem);
    let mut used_slugs: HashMap<String, usize> = HashMap::new();
    let mut emitted = 0;
    let mut skipped = 0;
    let mut collisions: Vec<String> = Vec::new();

    for section in &sections {
        if section.attributes.iter().any(|attribute| attribute.starts_with(":skip")) {
            skipped += 1;
            continue;
        }

        let expected_text = section.expected.join("\n");
        let is_error = expected_text.contains("(ERROR")
            || expected_text.contains("(MISSING")
            || section.attributes.iter().any(|attribute| attribute.starts_with(":error"));

        let mut slug = slugify(&section.name);
        if is_error {
            slug = format!("{slug}_ts_error");
        }
        let count = used_slugs.get(&slug).copied().unwrap_or(0) + 1;
        used_slugs.insert(slug.clone(), count);
        if count > 1 {
            slug = format!("{slug}_{count}");
        }

        for source_line in &section.source {
            if source_line.starts_with("#====")
                || source_line.starts_with("#----")
                || source_line.starts_with("#++++")
            {
                collisions.push(format!("{stem}/{slug}: {}", python_repr(source_line)));
            }
        }

        fixture_lines.push(format!("#---- {slug}"));
        fixture_lines.extend(section.source.iter().cloned());
        fixture_lines.push("#++++".to_string());
        fixture_lines.push(String::new());

        fs::create_dir_all(&expected_stem_dir)?;
        fs::write(
            expected_stem_dir.join(format!("{slug}.sexp")),
            expected_text + "\n",
        )?;
        emitted += 1;
    }

    fs::create_dir_all(fixture_dir)?;
    fs::write(
        fixture_dir.join(format!("{stem}.R.test")),
        fixture_lines.join("\n") + "\n",
    )?;

    if !collisions.is_empty() {
        println!(
            "  WARNING: {} source line(s) collide with fixture directives:",
            collisions.len()
        );
        for collision in &collisions {
            println!("    {collision}");
        }
    }

    Ok((emitted, skipped))
}

struct Section {
    name: String,
    attributes: Vec<String>,
    source: Vec<String>,
    expected: Vec<String>,
}

fn parse_sections(text: &str) -> Vec<Section> {
    let normalized = text.replace("\r\n", "\n").replace('\r', "\n");
    let lines: Vec<&str> = normalized.split('\n').collect();
    let mut sections = Vec::new();

    let mut index = 0;
    while index < lines.len() {
        let Some((close_index, content)) = try_header(&lines, index) else {
            // Stray content before the first header (or a blank line).
            index += 1;
            continue;
        };
        let name_parts: Vec<&str> = content
            .iter()
            .map(|line| line.trim())
            .filter(|line| !line.starts_with(':'))
            .collect();
        let name = name_parts.join(" ").trim().to_string();
        let attributes: Vec<String> = content
            .iter()
            .map(|line| line.trim())
            .filter(|line| line.starts_with(':'))
            .map(str::to_string)
            .collect();

        let body_start = close_index + 1;
        let mut next_header = lines.len();
        let mut scan = body_start;
        while scan < lines.len() {
            if try_header(&lines, scan).is_some() {
                next_header = scan;
                break;
            }
            scan += 1;
        }
        let body = &lines[body_start..next_header];

        let last_dash = body
            .iter()
            .enumerate()
            .filter(|(_, line)| is_dash(line))
            .map(|(position, _)| position)
            .next_back();
        let (source, expected): (&[&str], &[&str]) = match last_dash {
            Some(divider) => (&body[..divider], &body[divider + 1..]),
            None => (body, &[]),
        };

        sections.push(Section {
            name,
            attributes,
            source: strip_blank_edges(source),
            expected: strip_blank_edges(expected),
        });
        index = next_header;
    }

    sections
}

/// If a section header opens at `lines[index]`, return the index of its
/// closing `=` fence and the header content lines between the fences.
fn try_header<'lines>(lines: &[&'lines str], index: usize) -> Option<(usize, Vec<&'lines str>)> {
    if index >= lines.len() || !is_equals(lines[index]) {
        return None;
    }
    let mut cursor = index + 1;
    let mut content: Vec<&str> = Vec::new();
    while cursor < lines.len() && !is_equals(lines[cursor]) {
        content.push(lines[cursor]);
        cursor += 1;
    }
    if cursor >= lines.len() {
        return None; // no closing fence: not a header
    }
    if content.iter().all(|line| line.trim().is_empty()) {
        return None; // empty header block: not a header
    }
    Some((cursor, content))
}

fn is_fence(line: &str, fence: char) -> bool {
    let stripped = line.trim_end();
    stripped.len() >= 3 && stripped.chars().all(|character| character == fence)
}

fn is_equals(line: &str) -> bool {
    is_fence(line, '=')
}

fn is_dash(line: &str) -> bool {
    is_fence(line, '-')
}

fn slugify(name: &str) -> String {
    let mut slug = String::new();
    for character in name.chars() {
        if character.is_alphanumeric() {
            slug.extend(character.to_lowercase());
        } else {
            slug.push('_');
        }
    }
    let mut slug = slug.trim_matches('_').to_string();
    while slug.contains("__") {
        slug = slug.replace("__", "_");
    }
    if slug.is_empty() { "case".to_string() } else { slug }
}

/// Python's `repr()` for the collision warning: single quotes (double when the
/// text contains a single quote but no double quote), ASCII control escapes.
fn python_repr(text: &str) -> String {
    let quote = if text.contains('\'') && !text.contains('"') { '"' } else { '\'' };
    let mut repr = String::with_capacity(text.len() + 2);
    repr.push(quote);
    for character in text.chars() {
        match character {
            '\\' => repr.push_str("\\\\"),
            '\t' => repr.push_str("\\t"),
            '\n' => repr.push_str("\\n"),
            '\r' => repr.push_str("\\r"),
            character if character == quote => {
                repr.push('\\');
                repr.push(character);
            }
            character if character.is_ascii_control() => {
                repr.push_str(&format!("\\x{:02x}", character as u32));
            }
            character => repr.push(character),
        }
    }
    repr.push(quote);
    repr
}

fn strip_blank_edges(lines: &[&str]) -> Vec<String> {
    let mut start = 0;
    let mut end = lines.len();
    while start < end && lines[start].trim().is_empty() {
        start += 1;
    }
    while end > start && lines[end - 1].trim().is_empty() {
        end -= 1;
    }
    lines[start..end].iter().map(|line| line.to_string()).collect()
}

fn repo_root() -> Result<PathBuf, Box<dyn Error>> {
    let script_path = std::env::args().next().ok_or("cannot determine the script path")?;
    let script_dir = Path::new(&script_path)
        .parent()
        .map_or_else(|| PathBuf::from("."), Path::to_path_buf);
    Ok(fs::canonicalize(script_dir.join(".."))?)
}
